### ISSUE TITLE
Error When Computing the Pseudoinverse

### ISSUE THREAD

**Issue by wanzhe45** [tum-pbs/ConFIG#20]
Thank you for your excellent work. I am currently trying to apply your training method to my multi-task learning problem, which involves three loss functions.
I found that, due to the large number of network parameters in my task, the shape of grads is [3, 171798203]. When computing the pseudoinverse, both lstsq and pinv result in the following error:
torch._C._LinAlgError: cusolver error: CUSOLVER_STATUS_INVALID_VALUE, when calling cusolverDnSgesvdj_bufferSize(handle, jobz, econ, m, n, A, lda, S, U, ldu, V, ldv, lwork, params). This error may appear if the input matrix contains NaN. If you keep seeing this error, you may use torch.backends.cuda.preferred_linalg_library() to try linear algebra operators with other supported backends. See https://pytorch.org/docs/stable/backends.html#torch.backends.cuda.preferred_linalg_library
How can I solve this problem?
@qiauil

**Comment by qiauil** [comment]
Hi, @wanzhe45 

Thanks for using ConFIG. 

It seems that the error occurs because you have Nan values in your gradient matrix. This will happen when

1. You have some parameters unused in your network, and those parameters will have Nan values when you backpropagate through the network. To solve this, first check whether you really want to have these unused parameters in your network. If you do want to keep them,  you can change the "none_grad_mode" in `get_gradient_vector` function (https://tum-pbs.github.io/ConFIG/api/utils/#conflictfree.utils.get_gradient_vector).

2. If the Nan value is not due to the unused parameter. Then you should check whether your training hyperparameter is correct. This is usually because the loss is Nan at some iterations, which results in a Nan gradient.

Cheers,
Qiang

**Comment by wanzhe45** [comment]
Thank you for your reply. I checked my gradient matrix grads, and it does not contain any NaN values.

I suspect that the issue may still be caused by the large scale of my network. Specifically, the gradient matrix grads has a shape of [3, 171798203], which may be extremely ill-conditioned. Therefore, using PyTorch's lstsq or pinv to compute the pseudoinverse of this matrix results in the error mentioned above.

I then modified the pseudoinverse-related computation as follows:
```

if not isinstance(grads, torch.Tensor):
        grads = torch.stack(grads)
    with torch.no_grad():
        weights = weight_model.get_weights(
            gradients=grads, losses=losses, device=grads.device
        )
        # units = torch.nan_to_num((grads / (grads.norm(dim=1)).unsqueeze(1)), 0)
        if use_least_square:
            # best_direction = torch.linalg.lstsq(units, weights).solution
            best_direction = custom_lst_square(grads, weights)
        else:
            # best_direction = torch.linalg.pinv(units) @ weights
            best_direction = custom_pseudo_inverse(grads, weights)
        return length_model.rescale_length(
            target_vector=best_direction,
            gradients=grads,
            losses=losses,
        )
```
```
 def custom_lst_square(A: torch.Tensor, b: torch.Tensor) -> torch.Tensor:
    """
    Custom least square solution for the equation Ax = b.

    Args:
        A (torch.Tensor): The matrix A.
        b (torch.Tensor): The vector b.
    Returns:
        torch.Tensor: The least square solution x.
    """
    K = A.shape[0]
    eps = 1e-12
    norms = torch.linalg.vector_norm(A, dim=1).clamp_min(eps)
    G = (A @ A.T) / (norms[:, None] * norms[None, :])
    G = 0.5 * (G + G.T)
    ridge = 1e-4
    I = torch.eye(K, device=A.device, dtype=A.dtype)
    alpha = torch.linalg.solve(G + ridge * I, b)
    return A.T @ (alpha / norms)
```
After this modification, the error no longer occurs.

**Comment by qiauil** [comment]
Hi, @wanzhe45 

Thanks for your feedback! It looks like you have a more stable implementation of least squares.

I will temporarily keep this issue open and take a closer look at your new implementation to see whether it is a good idea to add your implementation to the ConFIG library.
